### PR TITLE
NP Likeness with confidence value

### PR DESCRIPTION
**Pull request by apahl**
added a scoring variant with an additional confidence value between 0..1
that describes how many fragments of the tested compound were found in the model data set.

### CODE REVIEW

File: Contrib/NP_Score/npscorer.py [hunk 1]
```diff
@@ -21,56 +21,70 @@
...
-  if mol is None:
-    raise ValueError('invalid molecule')
-  fp = rdMolDescriptors.GetMorganFingerprint(mol, 2)
-  bits = fp.GetNonzeroElements()
+def scoreMolWConfidence(mol, fscore):
+    """Next to the NP Likeness Score, this function outputs a confidence value
+    between 0..1 that descibes how many fragments of the tested molecule
+    were found in the model data set (1: all fragments were found).
+
+    Returns tuple(NP Likeness, Confidnce)"""
```
Comment (gedeck) [inline Contrib/NP_Score/npscorer.py]: Use a `namedtuple` to return annotated results. 

Typo: Confidnce => Confidence

File: Contrib/NP_Score/npscorer.py [hunk 2]
```diff
@@ -21,56 +21,70 @@
...
+            score += fscore[bit]
 
-  # preventing score explosion for exotic molecules
-  if score > 4:
-    score = 4. + math.log10(score - 4. + 1.)
-  if score < -4:
-    score = -4. - math.log10(-4. - score + 1.)
-  return score
+    score /= float(mol.GetNumAtoms())
+    confidence = round(float(bits_found / len(bits)), 3)
```
Comment (gedeck) [inline Contrib/NP_Score/npscorer.py]: I wouldn't round the results. While I understand your intention, it won't help when printing. See for example:
https://stackoverflow.com/questions/455612/limiting-floats-to-two-decimal-points

File: Contrib/NP_Score/npscorer.py [hunk 3]
```diff
@@ -21,56 +21,70 @@
...
 
+    # preventing score explosion for exotic molecules
+    if score > 4:
+        score = 4. + math.log10(score - 4. + 1.)
+    if score < -4:
+        score = -4. - math.log10(-4. - score + 1.)
+    return score, confidence
+
+
+def scoreMol(mol, fscore):
```
Comment (gedeck) [inline Contrib/NP_Score/npscorer.py]: As you already started to add a comment above, why not add comment for the other methods too.

File: Contrib/NP_Score/npscorer.py [hunk 4]
```diff
@@ -21,56 +21,70 @@
...
-  if score < -4:
-    score = -4. - math.log10(-4. - score + 1.)
-  return score
+    score /= float(mol.GetNumAtoms())
+    confidence = round(float(bits_found / len(bits)), 3)
 
+    # preventing score explosion for exotic molecules
+    if score > 4:
+        score = 4. + math.log10(score - 4. + 1.)
+    if score < -4:
```
Comment (gedeck) [inline Contrib/NP_Score/npscorer.py]: This could be a `elif`

File: Contrib/NP_Score/npscorer.py [hunk 5]
```diff
@@ -21,56 +21,70 @@
...
+    return scoreMolWConfidence(mol, fscore)[0]
 
 def processMols(fscore, suppl):
-  sys.stderr.write("calculating ...\n")
-  count = {}
-  n = 0
-  for i, m in enumerate(suppl):
-    if m is None:
-      continue
+    sys.stderr.write("calculating ...\n")
```
Comment (gedeck) [inline Contrib/NP_Score/npscorer.py]: Could use instead:
```
  print('calculating ...', file=sys.stderr)
```

File: Contrib/NP_Score/npscorer.py [hunk 6]
```diff
@@ -21,56 +21,70 @@
...
+        score = 4. + math.log10(score - 4. + 1.)
+    if score < -4:
+        score = -4. - math.log10(-4. - score + 1.)
+    return score, confidence
+
+
+def scoreMol(mol, fscore):
+    return scoreMolWConfidence(mol, fscore)[0]
 
 def processMols(fscore, suppl):
```
Comment (gedeck) [inline Contrib/NP_Score/npscorer.py]: You could add a keyword arg to optionally output the confidence score here.

Comment (apahl) [reply Contrib/NP_Score/npscorer.py]: I dont't like it when function options change the return type.

Comment (gedeck) [reply Contrib/NP_Score/npscorer.py]: My suggestion would be an extension of the method without changing the default behaviour.

File: Contrib/NP_Score/npscorer.py (hunk 1 above)
Comment (gedeck) [reply Contrib/NP_Score/npscorer.py]: Do you want to comment on the suggestion to use `namedtuple`? It still behaves like a tuple, but also allows access to individual fields by name.
```
NPLikeness = namedtuple('NPLikeness', 'npLikeness,confidence')

return NPLikeness(likeness, confidence)
```
and then you can use it like this:
```
score = scoreMolWConfidence(mol, fscore)
score.npLikeness == score[0]
score.confidence == score[1]
```

Comment (apahl) [reply Contrib/NP_Score/npscorer.py]: Personally I don't like Namedtuples too much, but it is probably a good suggestion. Will add a PR tomorrow.

File: Contrib/NP_Score/npscorer.py (hunk 4 above)
Comment (apahl) [reply Contrib/NP_Score/npscorer.py]: Didn't want to change all the legacy code... :-P

File: Contrib/NP_Score/npscorer.py (hunk 5 above)
Comment (apahl) [reply Contrib/NP_Score/npscorer.py]: Again, this is from the original code that I didn't want to change.

File: Contrib/NP_Score/npscorer.py (hunk 2 above)
Comment (apahl) [reply Contrib/NP_Score/npscorer.py]: On my machine (Python 3.5.3) this works:
```Python
>>> a = 13.949999999999999
>>> round(a, 2)
13.95
```

**Comment by greglandrum** [comment]
Thanks Axel!

**Comment by greglandrum** [comment]
Thanks Axel and Peter (for the review)!

**Comment by greglandrum** [comment]
Thanks Axel and Peter (for the review)!

**Pull request merged**